src/tree/kat_encryption.rs

```diff
@@ -0,0 +1,425 @@
+            assert_eq!(leaf.generations, leaf.sender_data_info.len() as u32);
+            let leaf_index = LeafIndex::from(leaf_index);
+
+            for (generation, application_key, handshake_key, message, sender_data_info) in izip!(
+                (0..leaf.generations).into_iter(),
+                &leaf.application_keys,
+                &leaf.handshake_keys,
+                &leaf.messages,
+                &leaf.sender_data_info
+            ) {
+                // Setup group
+                let mls_ciphertext = MLSCiphertext::decode(&mut Cursor::new(&message.ciphertext))
+                    .expect("Error parsing MLSCiphertext");
+                let mut group = receiver_group(ciphersuite, &mls_ciphertext.group_id);
+                *group.epoch_secrets_mut().sender_data_secret_mut() =
+                    SenderDataSecret::from(test_vector.sender_data_secret.as_slice());
+
+                // Check keys
+                let (application_secret_key, application_secret_nonce) = secret_tree
+                    .secret_for_decryption(
+                        ciphersuite,
+                        leaf_index,
+                        SecretType::ApplicationSecret,
+                        generation,
+                    )
+                    .expect("Error getting decryption secret");
+                assert_eq!(application_key.key, application_secret_key.as_slice());
+                assert_eq!(application_key.nonce, application_secret_nonce.as_slice());
+
+                let (handshake_secret_key, handshake_secret_nonce) = secret_tree
+                    .secret_for_decryption(
+                        ciphersuite,
+                        leaf_index,
+                        SecretType::HandshakeSecret,
+                        generation,
+                    )
+                    .expect("Error getting decryption secret");
+                assert_eq!(handshake_key.key, handshake_secret_key.as_slice());
+                assert_eq!(handshake_key.nonce, handshake_secret_nonce.as_slice());
+
+                let sender_data_key = AeadKey::from_sender_data_secret(
+                    ciphersuite,
+                    &sender_data_info.ciphertext,
+                    &sender_data_secret,
+                );
+                let sender_data_nonce = AeadNonce::from_sender_data_secret(
+                    ciphersuite,
+                    &sender_data_info.ciphertext,
+                    &sender_data_secret,
+                );
+                assert_eq!(sender_data_info.secrets.key, sender_data_key.as_slice());
+                assert_eq!(sender_data_info.secrets.nonce, sender_data_nonce.as_slice());
+
+                // Decrypt and check message
+                let indexed_members = HashMap::new();
+                let epoch_secret = group.epoch_secrets();
+                let context = group.context();
+
+                // The decryption doesn't check the plaintext in test mode!
```

Comment:
What do you mean by check? Is this alluding to the early return introduced earlier, that skips the last assert?

Reply:
Yep, i.e. the decryption here is all fine but the sender doesn't exist, hence the decryption should fail in the openmls logic. But we really want to test encryption/decryption here. After refactoring there should be a decryption function on the tree that we can call here.